### ISSUE TITLE
Point Cloud/camera extrinsics and rasterizer frame of reference mismatch

### ISSUE THREAD

**Issue by gaperezsa** [graphdeco-inria/diff-gaussian-rasterization#52]
I'm having trouble matching the frame of reference between the extrinsic of my cameras, the world point cloud and the rasterizers interpretation of the camera views. and i saw your email at the top of said file and wondered if you could help me.

Im attaching the initial rendering result from the first training epoch. At this point, the Gaussians were initialized with a SFM pointcloud and all data is in COLMAP format. It looks like:
![QuestionRender](https://github.com/graphdeco-inria/diff-gaussian-rasterization/assets/56688374/830baa88-ce4d-4423-a926-0e80116f7346)

Some debugging info from this render can be found here:
(Pdb) p cam_type
'colmap'
(Pdb) p viewpoint_camera.R
array([[ 0.43241069,  0.21258201, -0.87625903],
       [-0.51807059,  0.85396278, -0.04848126],
       [ 0.73798635,  0.47492785,  0.47939512]])
(Pdb) p viewpoint_camera.T
array([ 0.07173432, -0.20056451, -0.00144203])
(Pdb) p viewpoint_camera.FoVx
1.7121016721421645
(Pdb) p viewpoint_camera.FoVy
1.7121016721421645
(Pdb) p viewpoint_camera.image_height
1400
(Pdb) p viewpoint_camera.image_width
1400
(Pdb) p scaling_modifier
1.0
(Pdb) p viewpoint_camera.world_view_transform
tensor([[ 0.4324,  0.2126, -0.8763,  0.0000],
        [-0.5181,  0.8540, -0.0485,  0.0000],
        [ 0.7380,  0.4749,  0.4794,  0.0000],
        [ 0.0717, -0.2006, -0.0014,  1.0000]])
(Pdb) p viewpoint_camera.full_proj_transform
tensor([[ 0.3753,  0.1845, -0.8763, -0.8763],
        [-0.4496,  0.7411, -0.0485, -0.0485],
        [ 0.6404,  0.4121,  0.4794,  0.4794],
        [ 0.0623, -0.1741, -0.0114, -0.0014]])
(Pdb) p pc.active_sh_degree
0
(Pdb) p viewpoint_camera.camera_center
tensor([ 0.0717, -0.2006, -0.0014])
(Pdb) p means3D
Parameter containing:
tensor([[-0.1647, -0.3434, -0.5982],
        [-0.3601, -0.5212, -0.6345],
        [-0.7570, -1.4368, -1.3412],
        ...,
        [-2.6000,  0.6584,  0.3351],
        [-2.3827, -0.7138,  0.7575],
        [-0.5284, -0.1488, -0.6624]], device='cuda:0', requires_grad=True)
(Pdb) p means2D
tensor([[0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.],
        ...,
        [0., 0., 0.],
        [0., 0., 0.],
        [0., 0., 0.]], device='cuda:0')
(Pdb) p shs
tensor([[[-0.3823,  0.7159, -0.1321],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        [[ 1.5222,  0.5352,  1.3693],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        [[-0.7437,  1.2442, -0.3823],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        ...,

        [[ 0.1738, -1.0913,  1.2442],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        [[ 0.0626,  1.0774,  0.6881],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]],

        [[ 0.4240, -1.7586, -0.7993],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         ...,
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000],
         [ 0.0000,  0.0000,  0.0000]]], device='cuda:0')
(Pdb) p colors_precomp
None
(Pdb) p opacity
tensor([[0.1000],
        [0.1000],
        [0.1000],
        ...,
        [0.1000],
        [0.1000],
        [0.1000]], device='cuda:0')
(Pdb) p scales
Parameter containing:
tensor([[-4.1575, -4.1575, -4.1575],
        [-4.0598, -4.0598, -4.0598],
        [-0.6831, -0.6831, -0.6831],
        ...,
        [-2.5938, -2.5938, -2.5938],
        [-3.3329, -3.3329, -3.3329],
        [-4.1158, -4.1158, -4.1158]], device='cuda:0', requires_grad=True)
(Pdb) p rotations
Parameter containing:
tensor([[1., 0., 0., 0.],
        [1., 0., 0., 0.],
        [1., 0., 0., 0.],
        ...,
        [1., 0., 0., 0.],
        [1., 0., 0., 0.],
        [1., 0., 0., 0.]], device='cuda:0', requires_grad=True)
(Pdb) p cov3D_precomp
None
(Pdb) p rotations.shape
torch.Size([9149, 4])

As you can see, the render does not match with the cameras corresponding image. However, if i were to debug this uploading the point cloud to MeshLab and then using the extrinsics of this camera viewpoint to see the point cloud, then it looks like: 
![MeshLabDebuggingScreenshot02](https://github.com/graphdeco-inria/diff-gaussian-rasterization/assets/56688374/196494f8-cbdd-456a-bcc4-a77c69d9605f)
 , which matches the RGB image for that viewpoint quite nicely.

As an extra debugging step, in order to make sense of the first render by the rasterizer, I initialized with the point cloud and asked for it to render standing from 0,0,0 and using a RotationMatrix="0 0 -1 0 0 1 0 0 1 0 0 0 0 0 0 1 ", that is, a 270° y-axis rotation. That looks like: 
![GaussianRenderedDebuggingScreenshotYaxis270°03](https://github.com/graphdeco-inria/diff-gaussian-rasterization/assets/56688374/4f6f1e6c-01cd-4629-9fcd-a0e1af9b3f5e)

but, again, if i were to input that extrinsic in meshlab, i see: 
![MeshLabDebuggingScreenshotYaxis270°03](https://github.com/graphdeco-inria/diff-gaussian-rasterization/assets/56688374/6a0027c2-c5c6-43a6-8e53-eb6905bc5fa0)
. So there is definitely some rotation or swapping or inversion of axis between the coordinate system used for the point cloud/extrinsic and the rasterizer.

Any insight is well appreciated!
